### PR TITLE
IGNITE-14893 Fix IndexOutOfBoundsEcxeption in GridCacheWriteBehindStore flushers lookup by key

### PR DESCRIPTION
**Pull request by solveme**
Introduce several methods in IgniteUtils class to perform mapping hash code of arbitrary object to array indexes.
Apply introduced util methods instead of broken WB store method

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

**Comment by torahmike** [comment]
Why is there different logic when the thread count is a power of 2?

**Comment by solveme** [comment]
> Why is there different logic when the thread count is a power of 2?

This approach was taken from HashMap.hash(Object) logic. I guess originally it was chosen for a little better performance.

**Comment by solveme** [comment]
> Why is there different logic when the thread count is a power of 2?

This approach was taken from HashMap hashing logic. I guess originally it was chosen for a little better performance. Note that I only extracted this code from GridCacheWriteBehindStore and apply required fixes for negative hash code cases.

File: modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java [hunk 1]
```diff
@@ -12106,4 +12106,33 @@ public static long uncompressedSize(File zip) throws IOException {
...
+    }
+
+    /**
+     * Map object hash to some index between 0 and specified size using
+     * approach similar to {@link HashMap#hash(Object)}. In case
+     * <code>size</code> is not a power of 2 we fallback to modulo operation.
+     *
+     * @param hash Object hash.
+     * @param size Size greater than 0.
+     * @return Calculated index in range [0..size).
```
Comment (sk0x50) [inline modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java:12128]: Please add to Javadoc a new parameter which is `isSizePow2`.

File: modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java [hunk 2]
```diff
@@ -12106,4 +12106,33 @@ public static long uncompressedSize(File zip) throws IOException {
...
+     */
+    public static int hashToIndexOpt(int hash, int size, boolean isSizePow2) {
```
Comment (sk0x50) [inline modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java:12130]: Well, I don't think that you need to provide a unique name here, overloading is enough.

File: modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java [hunk 3]
```diff
@@ -12106,4 +12106,33 @@ public static long uncompressedSize(File zip) throws IOException {
...
+        int h = (hash ^ (hash >>> 16));
+
+        return isSizePow2
```
Comment (sk0x50) [inline modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java:12133]: Well, this implementation relies on the user input and does not explicitly check the fact that the size variable is a power of two.
It seems to me, it is easy to write a code as follows: hashToIndex(hash, 9, true) that leads to that only two results can be returned in this case - 0 and 8.

**Pull request closed without merging**